lib: propagate aborted state to dependent signals before firing events

jazelly opened the pull request:
Fixes: https://github.com/nodejs/node/issues/54466
Fixes: https://github.com/nodejs/node/issues/54601

jazelly commented:
I am not sure if I should update WPT together with this PR, considering @RedYetiDev has [drafted a PR](https://github.com/nodejs/node/pull/54468) to cover this.

RedYetiDev commented:
> I am not sure if I should update WPT together with this PR, considering @RedYetiDev has [drafted a PR](https://github.com/nodejs/node/pull/54468) to cover this.

No worries, once this lands, I’ll rebase it into my PR.

jazelly commented:
sure, I'll add a test

RedYetiDev commented on lib/internal/abort_controller.js:
```diff
@@ -373,17 +374,36 @@ ObjectDefineProperty(AbortSignal.prototype, SymbolToStringTag, {
+  // 3. Let dependentSignalsToAbort be a new list.
+  const dependentSignalsToAbort = [];
+  // 4. For each dependentSignal of signal's dependent signals:
+  signal[kDependantSignals]?.forEach((s) => {
```
Does this need a primordial?

RedYetiDev commented on lib/internal/abort_controller.js:
```diff
@@ -373,17 +374,36 @@ ObjectDefineProperty(AbortSignal.prototype, SymbolToStringTag, {
 defineEventHandler(AbortSignal.prototype, 'abort');
 
 function abortSignal(signal, reason) {
+  // 1. If signal is aborted, then return.
```
```suggestion
  // https://dom.spec.whatwg.org/#dom-abortsignal-abort
  // 1. If signal is aborted, then return.
```

RedYetiDev commented on lib/internal/abort_controller.js:
```diff
@@ -372,18 +374,51 @@ ObjectDefineProperty(AbortSignal.prototype, SymbolToStringTag, {
+
+// To run the abort steps for an AbortSignal signal
+function runAbort(signal) {
+  // TODO: 1. For each algorithm of signal's abort algorithms: run algorithm.
+  // TODO: 2. Empty signal's abort algorithms.
```
```suggestion
  // TODO(jazelly): 1. For each algorithm of signal's abort algorithms: run algorithm.
  // TODO(jazelly): 2. Empty signal's abort algorithms.
```

When adding TODO comments, putting the username of who added it can help with organization.

RedYetiDev commented on test/parallel/test-abortsignal-any.mjs:
```diff
@@ -118,4 +118,43 @@ describe('AbortSignal.any()', { concurrency: !process.env.TEST_PARALLEL }, () =>
     controller.abort();
     assert.strictEqual(result, 1);
   });
+
```
Can you add the source? (Assuming this was taken from the WPT)

If not, my bad!

RedYetiDev commented:
Nice!

CC @shaseley (from the spec proposal)

jasnell commented on lib/internal/abort_controller.js:
```diff
@@ -372,18 +374,50 @@ ObjectDefineProperty(AbortSignal.prototype, SymbolToStringTag, {
+  const dependentSignalsToAbort = [];
+  // 4. For each dependentSignal of signal's dependent signals:
+  if (signal[kDependantSignals]) {
+    SetPrototypeForEach(signal[kDependantSignals], (s) => {
```
nit: I've always been a bit unhappy with the naming for this... Would be please consider adding a comment here indicating that `SetPrototypeForEach` is `Set.prototype.forEach` as opposed to "setting the prototype for each ..."

jazelly replied:
Haha, indeed. I'll add a comment

jasnell commented on lib/internal/abort_controller.js:
```diff
@@ -372,18 +374,50 @@ ObjectDefineProperty(AbortSignal.prototype, SymbolToStringTag, {
+// To run the abort steps for an AbortSignal signal
+function runAbort(signal) {
+  // TODO(jazelly): 1. For each algorithm of signal's abort algorithms: run algorithm.
+  // TODO(jazelly): 2. Empty signal's abort algorithms.
```
It's not immediately clear if these todo's are still relevant? Should they still be here or are they left over from development?

jazelly replied:
I'll remove these comments. These todo's are on my list, but probably worth it to bring it in a discussion first.

RedYetiDev commented:
@jasnell Is this semver-minor, because it changes an existing behavior (where the existing behavior won't cause breakages because it's an internal error)

jasnell commented:
Probably safest to mark it as such

jazelly commented:
looks like I need another approval after addressing the PR comments.

jakecastelli reviewed (commented):
Looks like this PR has merge conflicts now

aduh95 commented:
> this should be ready for an author to 🚀 land it.

The only author in the PR is jazelly. https://github.com/nodejs/node/labels/author%20ready does not mean the PR should be merged by its author, rather than the author has nothing more to do.

aduh95 commented on lib/internal/abort_controller.js:
```diff
@@ -372,18 +374,48 @@ ObjectDefineProperty(AbortSignal.prototype, SymbolToStringTag, {
+  // 3. Let dependentSignalsToAbort be a new list.
+  const dependentSignalsToAbort = [];
+  // 4. For each dependentSignal of signal's dependent signals:
+  if (signal[kDependantSignals]) {
+    // Equivalent to Set.prototype.forEach
+    SetPrototypeForEach(signal[kDependantSignals], (s) => {
```
We're using a `SafeSet`, there's no point in using `SetPrototypeForEach`
```suggestion
  if (signal[kDependantSignals]) {
    signal[kDependantSignals].forEach((s) => {
```
What about using optional chaining btw?
```suggestion
  signal[kDependantSignals]?.forEach((s) => {
```

jazelly replied:
makes sense. If primordials is unnecessary then optional chaining is better.

aduh95 commented on lib/internal/abort_controller.js:
```diff
@@ -372,18 +374,48 @@ ObjectDefineProperty(AbortSignal.prototype, SymbolToStringTag, {
   signal[kAborted] = true;
   signal[kReason] = reason;
+  // 3. Let dependentSignalsToAbort be a new list.
+  const dependentSignalsToAbort = [];
```
nit: we could use a null-prototype array to avoid prototype pollution here.
```suggestion
  const dependentSignalsToAbort = ObjectSetPrototypeOf([], null);
```

jakecastelli commented:
Hi folks, I was wondering if it makes sense to let the PR author also update the WPT related tests in the same PR (in separate commits), I do appreciate the effort of doing so in 54468 @RedYetiDev, on the other hand I feel like it may be more consistent to have them in one PR (easy to know if any relate test would fail or not) plus for future debugging / tracing purpose.
cc. @jazelly @aduh95 @jasnell

RedYetiDev commented:
SGTM, I'm happy to have my PR just take care of the leftovers (that haven't been updated from PRs like this)

jazelly commented:
Awesome. I have updated the PR with latest wpt for `abort` with a separate commit. Also addressed PR comments and redundant test cases in the original commit. PTAL

jasnell commented:
Landed in d47360604047
